src/backend/marsha/core/api/video.py

```diff
@@ -465,6 +478,116 @@ def update_live_state(
+            & (permissions.IsTokenInstructor | permissions.IsTokenAdmin)
+        ],
+    )
+    # pylint: disable=unused-argument
+    def pairing_secret(
+        self,
+        request,
+        pk=None,
+    ):
+        """Generate a secret for pairing an external device to a live stream.
+
+        Deletes expired LivePairing objects.
+
+        Parameters
+        ----------
+        request : Type[django.http.request.HttpRequest]
+            The request on the API endpoint
+        pk: string
+            The primary key of the video
+
+        Returns
+        -------
+        Type[rest_framework.response.Response]
+            HttpResponse with the generated secret.
+        """
+        LivePairing.objects.delete_expired()
+
+        video = self.get_object()
+        if video.live_type != JITSI:
+            return Response(
+                {"detail": "Matching video is not a Jitsi Live."}, status=400
+            )
+
+        try:
+            live_pairing = LivePairing.objects.get(video=video)
+        except LivePairing.DoesNotExist:
+            live_pairing = LivePairing(video=video)
+
+        for _ in range(2):
+            try:
+                live_pairing.generate_secret()
+                live_pairing.save()
+                break
+            except ValidationError:
+                pass
+
+        serializer = serializers.LivePairingSerializer(instance=live_pairing)
+        return Response(serializer.data)
+
+
+class PairingChallengeThrottle(AnonRateThrottle):
+    """Throttling for pairing challenge requests."""
+
+    def get_ident(self, request):
+        """Identify requests by box_id instead of IP"""
+        try:
+            box_id = request.data.get("box_id")
+            Device.objects.get(pk=box_id)
+            return box_id
+        except (Device.DoesNotExist, ValidationError):
+            return None
+
+
+@api_view(["POST"])
+@throttle_classes([PairingChallengeThrottle])
+def pairing_challenge(request):
+    """View handling pairing challenge request to stream from an external device.
+
+    Deletes expired LivePairing objects.
+
+    Parameters
+    ----------
+    request : Type[django.http.request.HttpRequest]
+        The request on the API endpoint, it must contains a payload with the following fields:
+            - box_id: uuid from external device.
+            - secret: generated pairing secret.
+
+    Returns
+    -------
+    Type[rest_framework.response.Response]
+        HttpResponse containing live video credentials.
+
+    """
+    LivePairing.objects.delete_expired()
+
+    serializer = serializers.PairingChallengeSerializer(data=request.data)
+    if not serializer.is_valid():
+        return Response(serializer.errors, status=400)
+
+    try:
+        live_pairing = LivePairing.objects.get(secret=request.data.get("secret"))
+    except LivePairing.DoesNotExist:
+        return Response({"detail": "Secret not found."}, status=404)
+
+    if live_pairing.video.live_type != JITSI:
```

Comment:
If it's not a `JITSI` video, the `live_pairing` should be deleted then ?

Reply:
Indeed!

Reply:
So I think deleting the `live_pairing` should be moved just before this `if`. WDYT ?

Reply:
That's how I managed this, yes.